sound/soc/sof/ipc4-topology.c

```diff
@@ -1064,20 +1065,45 @@ sof_ipc4_prepare_copier_module(struct snd_sof_widget *swidget,
 		case SOF_DAI_INTEL_ALH:
 		{
 			struct sof_ipc4_alh_configuration_blob *blob;
+			struct sof_ipc4_copier_data *alh_data;
+			struct sof_ipc4_copier *alh_copier;
+			struct snd_sof_widget *w;
 			u32 ch_map;
 			int i;
+			int j;
 
 			blob = (struct sof_ipc4_alh_configuration_blob *)ipc4_copier->copier_config;
-			blob->alh_cfg.count = 1;
-			blob->alh_cfg.mapping[0].alh_id = copier_data->gtw_cfg.node_id;
+
 			blob->gw_attr.lp_buffer_alloc = 0;
 			copier_data->gtw_cfg.config_length = sizeof(*blob) >> 2;
 
+			blob->alh_cfg.count = 0;
+
+			/* Set each gtw_cfg.node_id to blob->alh_cfg.mapping[] for all widgets with the same stream name */
+			list_for_each_entry(w, &sdev->widget_list, list) {
+				if (w->widget->sname &&
+				    strcmp(w->widget->sname, swidget->widget->sname))
+					continue;
+
+				dai = w->private;
+				alh_copier = (struct sof_ipc4_copier *)dai->private;
+				alh_data = &alh_copier->data;
+
+				blob->alh_cfg.mapping[blob->alh_cfg.count].alh_id = alh_data->gtw_cfg.node_id;
```

Comment:
I am not following how each DAI can have a value for gtw_cfg.node_id, and the lines 1096-1097 we assign a gtw_cfg.node_id.

who sets the node_id for each DAI?

Reply:
> I am not following how each DAI can have a value for gtw_cfg.node_id, and the lines 1096-1097 we assign a gtw_cfg.node_id.
> 
> who sets the node_id for each DAI?

gtw_cfg is a part of sof_ipc4_copier_data, each copier widget has its own gtw_cfg. We assign gtw_cfg.node_id in sof_ipc4_dai_config(). The node_id is associated with ALH stream id.
```
sdw_params_stream -> hda_ctrl_dai_widget_setup -> sof_ipc4_dai_config
```